### PR TITLE
[Algorithm] IMPALA and VTrace module

### PR DESCRIPTION
**Pull request by albertbou92**
## Description

This PR adds a VTrace module to be able to run IMPALA  (https://arxiv.org/abs/1802.01561)

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: torchrl/objectives/value/vtrace.py [hunk 1]
```diff
@@ -4,55 +4,334 @@
...
-    for t in range(T - 2, -1, -1):
-        _v_out = (
-            vals[:, t] + dv[:, t] + gamma[:, t] * c[:, t] * (v_out[-1] - vals[:, t + 1])
+    """Computes V-Trace off-policy actor critic targets.
+
+    Refer to "IMPALA: Scalable Distributed Deep-RL with Importance Weighted  Actor-Learner Architectures"
+    https://arxiv.org/abs/1802.01561 for more context.
+
+    Args:
+        gamma (scalar): exponential mean discount.
```
Comment (vmoens) [inline torchrl/objectives/value/vtrace.py:69]: will this work with gamma as a tensor (per-value gamma)?
If not we should raise an error saying that it is not supported yet and we welcome issues and contributions etc.

File: torchrl/objectives/value/vtrace.py [hunk 2]
```diff
@@ -4,55 +4,334 @@
...
+        log_pi (Tensor): log probability of taking actions in the environment.
+        log_mu (Tensor): log probability of taking actions in the environment.
+        state_value (Tensor): value function result with old_state input.
```
Comment (vmoens) [inline torchrl/objectives/value/vtrace.py:72]: what is `old_state` input?

File: torchrl/objectives/value/vtrace.py [hunk 3]
```diff
@@ -4,55 +4,334 @@
...
+        next_state_value (Tensor): value function result with new_state input.
```
Comment (vmoens) [inline torchrl/objectives/value/vtrace.py:73]: ditto with `next_state`

File: torchrl/objectives/value/vtrace.py [hunk 4]
```diff
@@ -4,55 +4,334 @@
...
+    device = state_value.device
+
+    deltas, clipped_rho = _dv_val(reward, state_value, next_state_value, gamma, rho_thresh, log_pi, log_mu)
+    clipped_c = torch.min(torch.tensor(c_thresh).to(device), clipped_rho)
+
+    ############################################################
+    # MAKE THIS PART WORK; THEN WE CAN TRY TO MAKE IT FASTER
+
+    not_done = (~done).int()
+    *batch_size, time_steps, lastdim = not_done.shape
```
Comment (vmoens) [inline torchrl/objectives/value/vtrace.py:98]: if you don't implement @_transpose_time this will break when people pass `time_dim`

File: examples/impala/impala.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
+"""
```
Comment (vmoens) [inline examples/impala/impala.py:1]: add headers

File: torchrl/objectives/value/vtrace.py [hunk 6]
```diff
@@ -4,55 +4,334 @@
...
-            vals[:, t] + dv[:, t] + gamma[:, t] * c[:, t] * (v_out[-1] - vals[:, t + 1])
+    """Computes V-Trace off-policy actor critic targets.
+
+    Refer to "IMPALA: Scalable Distributed Deep-RL with Importance Weighted  Actor-Learner Architectures"
+    https://arxiv.org/abs/1802.01561 for more context.
+
+    Args:
+        gamma (scalar): exponential mean discount.
+        log_pi (Tensor): log probability of taking actions in the environment.
+        log_mu (Tensor): log probability of taking actions in the environment.
```
[the remark this answers is not in the document]
Comment (albertbou92) [reply torchrl/objectives/value/vtrace.py:71]: one is the logprob of the collection actor, the other of the current actor. Added!

File: torchrl/objectives/value/vtrace.py (hunk 2 above)
Comment (albertbou92) [reply torchrl/objectives/value/vtrace.py:72]: actually it is just state. I will clarify.

File: test/test_cost.py [hunk 7]
```diff
@@ -1064,7 +1052,6 @@ def test_qmix_tensordict_keys(self, td_est):
             "action": ("agents", "action"),
             "reward": "reward",
             "done": "done",
-            "terminated": "terminated",
         }
```
Comment (vmoens) [inline test/test_cost.py:1055]: Why are all these `"terminated"` gone? (not copying the same comment everywhere but I see most of them have disappeared)

File: torchrl/objectives/value/advantages.py [hunk 8]
```diff
@@ -1328,6 +1350,245 @@ def value_estimate(
...
+        Examples:
+            >>> from tensordict import TensorDict
+            >>> value_net = TensorDictModule(
+            ...     nn.Linear(3, 1), in_keys=["obs"], out_keys=["state_value"]
+            ... )
+            >>> module = VTrace(
+            ...     gamma=0.98,
+            ...     value_network=value_net,
+            ...     differentiable=False,
+            ... )
```
Comment (vmoens) [inline torchrl/objectives/value/advantages.py:1489]: does this run (without actor)?

File: torchrl/objectives/value/functional.py [hunk 9]
```diff
@@ -1211,6 +1212,78 @@ def vec_td_lambda_advantage_estimate(
...
+    ``[*Batch x TimeSteps x *F]``, with ``*F`` feature dimensions.
+    """
+    if not (next_state_value.shape == state_value.shape == reward.shape == done.shape):
+        raise RuntimeError(SHAPE_ERR)
+
+    device = state_value.device
+
+    not_done = (~done).int()
+    *batch_size, time_steps, lastdim = not_done.shape
+    discounts = gamma * not_done
```
Comment (vmoens) [inline torchrl/objectives/value/functional.py:1260]:
```suggestion
    discounts_deltas = gamma * not_terminated
    discounts = gamma * not_done
```

File: torchrl/objectives/value/functional.py [hunk 10]
```diff
@@ -1211,6 +1212,78 @@ def vec_td_lambda_advantage_estimate(
...
+
+    rho = (log_pi - log_mu).exp()
+    clipped_rho = rho.clamp_max(rho_thresh)
+    deltas = clipped_rho * (reward + discounts * next_state_value - state_value)
```
Comment (vmoens) [inline torchrl/objectives/value/functional.py:1264]:
```suggestion
    deltas = clipped_rho * (reward + discounts_deltas * next_state_value - state_value)
```

File: torchrl/objectives/value/functional.py [hunk 11]
```diff
@@ -1211,6 +1212,78 @@ def vec_td_lambda_advantage_estimate(
...
+
+    All tensors (values, reward and done) must have shape
+    ``[*Batch x TimeSteps x *F]``, with ``*F`` feature dimensions.
+    """
+    if not (next_state_value.shape == state_value.shape == reward.shape == done.shape):
+        raise RuntimeError(SHAPE_ERR)
+
+    device = state_value.device
+
+    not_done = (~done).int()
```
Comment (vmoens) [inline torchrl/objectives/value/functional.py:1258]:
```suggestion
    not_done = (~done).int()
    not_terminated = (~terminated).int()
```

**Review by vmoens — COMMENTED** [review]
Amazing stuff! We wanted to see that from the early days of the lib, I'm super excited by this!
In general these changes all make sense to me.

The test_cost seems to be going backward a bit, loads of terminated have disappeared. Can we merge them back?

Have you tried to see if the vtrace function gets faster when you decorate it with torch.compile?

We need to integrate terminated in the vtrace function but that should be easy. Once it's done you can copy paste the logic of other advantage module to have that functionality in VTrace too!

File: examples/impala/utils.py [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+# LICENSE file in the root directory of this source tree.
+
+import os
+
+# To pickle the environment, in particular the EndOfLifeTransform, we need to
+# add the utils path to the PYTHONPATH
+utils_path = os.path.abspath(os.path.abspath(os.path.dirname(__file__)))
+current_pythonpath = os.environ.get("PYTHONPATH", "")
+new_pythonpath = f"{utils_path}:{current_pythonpath}"
+os.environ["PYTHONPATH"] = new_pythonpath
```
[the remark this answers is not in the document]
Comment (albertbou92) [reply examples/impala/utils.py:13]: We can place it after imports. Regarding whether or not should be a proper transform, I have doubts. Is too specific for Atari, it is not a general Transform

File: torchrl/objectives/value/advantages.py (hunk 8 above)
Comment (albertbou92) [reply torchrl/objectives/value/advantages.py:1489]: No, I will correct the docs

File: examples/impala/utils.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (vmoens) [reply examples/impala/utils.py]: Loads of repos have "atari" transforms
We can even put it in a separate file
I'm not super comfortable wit having a copy of a class across 4 different examples, it'll be hard to maintain on the long term (+ it's untested - hence quite dangerous)

File: test/test_cost.py [hunk 13]
```diff
@@ -10296,28 +10605,65 @@ def test_skip_existing(
...
-            shifted=shifted,
-            skip_existing=skip_existing,
-            **kwargs,
-        )
-        td = TensorDict(
-            {
-                "obs": torch.randn(1, 10, 3),
-                "state_value": torch.ones(1, 10, 1),
-                "next": {
+        if adv == VTrace:
```
Comment (vmoens) [inline test/test_cost.py:10608]: ditto

File: test/test_cost.py [hunk 14]
```diff
@@ -10335,15 +10681,34 @@ def test_skip_existing(
...
         ],
     )
     def test_set_keys(self, value, adv, kwargs):
         value_net = TensorDictModule(nn.Linear(3, 1), in_keys=["obs"], out_keys=[value])
-        module = adv(
-            gamma=0.98,
-            value_network=value_net,
-            **kwargs,
-        )
+        if adv == VTrace:
```
Comment (vmoens) [inline test/test_cost.py:10689]: ditto

File: test/test_cost.py [hunk 15]
```diff
@@ -10365,14 +10731,36 @@ def test_set_deprecated_keys(self, adv, kwargs):
...
-            module = adv(
-                gamma=0.98,
-                value_network=value_net,
-                value_key="test_value",
-                advantage_key="advantage_test",
-                value_target_key="value_target_test",
-                **kwargs,
-            )
+
+            if adv == VTrace:
```
Comment (vmoens) [inline test/test_cost.py:10735]: ditto

File: torchrl/objectives/value/advantages.py [hunk 16]
```diff
@@ -597,7 +619,7 @@ def value_estimate(
 
         if self.average_rewards:
             reward = reward - reward.mean()
-            reward = reward / reward.std().clamp_min(1e-4)
+            reward = reward / reward.std().clamp_min(1e-5)
```
Comment (vmoens) [inline torchrl/objectives/value/advantages.py:622]: is there a reason to change this?
These small changes can introduce unforseen effects in our benchmarks. If we need to adapt it for the purpose of this PR I guess we should use an arg or smth like that to control it

File: torchrl/objectives/value/advantages.py [hunk 17]
```diff
@@ -1328,6 +1350,285 @@ def value_estimate(
         return value_target
 
 
+class VTrace(ValueEstimatorBase):
+    """A class wrapper around V-Trace estimate functional.
+
+    Refer to "IMPALA: Scalable Distributed Deep-RL with Importance Weighted  Actor-Learner Architectures"
+    https://arxiv.org/abs/1802.01561 for more context.
```
Comment (vmoens) [inline torchrl/objectives/value/advantages.py:1357]: will this appear as a link? I'm not sure. If not we should use sphinx hlinks

File: torchrl/objectives/value/advantages.py [hunk 18]
```diff
@@ -1328,6 +1350,285 @@ def value_estimate(
...
+
+    Args:
+        gamma (scalar): exponential mean discount.
+        value_network (TensorDictModule): value operator used to retrieve the value estimates.
+        actor_network (TensorDictModule): actor operator used to retrieve the log prob.
+        rho_thresh (Union[float, Tensor]): rho clipping parameter for importance weights.
```
Comment (vmoens) [inline torchrl/objectives/value/advantages.py:1363]:
```suggestion
        rho_thresh (Union[float, Tensor]): rho clipping parameter for importance weights.
            Defaults to ``1.0``.
```

File: torchrl/objectives/value/advantages.py [hunk 19]
```diff
@@ -1328,6 +1350,285 @@ def value_estimate(
...
+        c_thresh (Union[float, Tensor]): c clipping parameter for importance weights.
```
Comment (vmoens) [inline torchrl/objectives/value/advantages.py:1364]:
```suggestion
        c_thresh (Union[float, Tensor]): c clipping parameter for importance weights.
            Defaults to ``1.0``.
```

File: torchrl/objectives/value/advantages.py [hunk 20]
```diff
@@ -1328,6 +1350,285 @@ def value_estimate(
...
+        actor_network: TensorDictModule,
+        value_network: TensorDictModule,
+        rho_thresh: Union[float, torch.Tensor] = 1.0,
+        c_thresh: Union[float, torch.Tensor] = 1.0,
+        average_adv: bool = False,
+        differentiable: bool = False,
+        skip_existing: Optional[bool] = None,
+        advantage_key: NestedKey = None,
+        value_target_key: NestedKey = None,
+        value_key: NestedKey = None,
```
Comment (vmoens) [inline torchrl/objectives/value/advantages.py:1416]:
```suggestion
        advantage_key: Optional[NestedKey] = None,
        value_target_key: Optional[NestedKey] = None,
        value_key: Optional[NestedKey] = None,
```

File: torchrl/objectives/value/advantages.py [hunk 21]
```diff
@@ -1328,6 +1350,285 @@ def value_estimate(
...
+            differentiable=differentiable,
+            advantage_key=advantage_key,
+            value_target_key=value_target_key,
+            value_key=value_key,
+            skip_existing=skip_existing,
+        )
+        try:
+            device = next(value_network.parameters()).device
+        except (AttributeError, StopIteration):
+            device = torch.device("cpu")
```
Comment (vmoens) [inline torchrl/objectives/value/advantages.py:1431]: we probably want to use the default device here, I think `None` should work just fine

File: torchrl/objectives/value/advantages.py [hunk 22]
```diff
@@ -1328,6 +1350,285 @@ def value_estimate(
...
+        extended_in_keys = parent_in_keys + [self.tensor_keys.sample_log_prob]
+        return extended_in_keys
+
+    @_self_set_skip_existing
+    @_self_set_grad_enabled
+    @dispatch
+    def forward(
+        self,
+        tensordict: TensorDictBase,
+        *unused_args,
```
Comment (vmoens) [inline torchrl/objectives/value/advantages.py:1463]: do we test what these guys are?

Comment (albertbou92) [reply torchrl/objectives/value/advantages.py:1465]: The unused_args? I added them because GAE has them, are they even necessary?

Comment (vmoens) [reply torchrl/objectives/value/advantages.py:1465]: I don't see them appearing after that, so I would suggest to do 
```python
    def forward(
        self,
        tensordict: TensorDictBase,
        *, ...
```
for this and for GAE.

**Review by vmoens — APPROVED** [review]
LGTM, can you edit the unused args?
